ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,172 @@
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	logger.Info(fmt.Sprintf("Processing %s with MessageId %s", TaskManifestMessageName, aws.StringValue(message.MessageId)))
+	m := tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration)
+	err := tmr.processTaskManifestMessage(message)
+	m.Done(err)
+}
+
+// processTaskManifestMessage processes a task manifest message from ACS. It verifies
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum <= agentLatestSeqNum {
+		logger.Warn(fmt.Sprintf("Discarding task manifest message. "+
+			"Sequence number from manifest is less than Agent latest sequence number."), logger.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		})
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Compare running tasks on instance with those in the manifest.
+	tasksToStop, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message.
+	tmr.ackTaskManifestMessage(message)
+
+	// Send the task stop verification message to ACS.
+	tmr.sendTaskStopVerification(message, tasksToStop)
```

Comment:
Is there any specific reason we changed these calls on L127 and L130 to be synchronous instead of the existing asynchronous behavior? Either way, I don't think it's a big deal since these calls are short lived, but just curious since they are currently set up as async calls such that they are not blocking.

Reply:
Just synced with the original author of the code who is just back from vacation. There is code history for them to be asynchronous, and ACS may also hang on the ack call for a bit so we want to send both at the same time.